src/DIRAC/WorkloadManagementSystem/DB/PilotAgentsDB.py

```diff
@@ -1050,3 +1142,81 @@ def getPilotMonitorWeb(self, selectDict, sortList, startItem, maxItems):
     resultDict['Records'] = records
 
     return S_OK(resultDict)
+
+
+class PivotedPilotSummaryTable:
+  """
+  The class creates a 'pivoted' table by combining records with with the same group
+  of self.columnList into a single row. It allows an easy calculateion of pilot efficiencies.
+  """
+
+  def __init__(self, columnList):
+    """
+    Initialise a table with columns to be grouped by.
+
+    :param columnList: i.e. ['GridSite', 'DestinationSite']
+    :return:
+    """
+
+    self.columnList = columnList
+    self.pstates = ['Submitted', 'Done', 'Failed', 'Aborted',
+                    'Running', 'Waiting', 'Scheduled', 'Ready']
+
+    # we want 'Site' and 'CE' in the final result
+    colMap = {'GridSite': 'Site', 'DestinationSite': 'CE'}
+    self.columns = [colMap.get(val, val) for val in columnList]
+
+    self.columns += self.pstates  # MySQL._query() does not give us column names, sadly.
+
+  def buildSQL(self, selectDict=None):
```

Comment:
This query creates a pretty lengthy and complex MySQL query. I am not against it but I would like to have a ready-to-use test in case one day we want to modify it.

Reply:
An integration test added.